test/Microsoft.ML.OnnxTransformerTest/OnnxTransformTests.cs

```diff
@@ -139,6 +139,19 @@ public void TestSimpleCase()
             }
             catch (ArgumentOutOfRangeException) { }
             catch (InvalidOperationException) { }
+
+            try
+            {
+                // MYTODO: Is this side effect acceptable? Should I remove this from this test?
+                // OnnxTransformer will drop all the columns in the input that have
+                // a corresponding input in the onnx model.
+                // A side effect of this, is that after applying a pretrained onnx model
+                // like the one in here, we won't be able to use or access the input
+                // columns of the model.
+                var transformedData = pipe.Fit(dataView).Transform(dataView);
+                var col = transformedData.Schema["data_0"];
```

Comment:
One side effect of dropping all the input columns that were used by the onnx model, is that, obviously, we won't be able to access those input columns.

In scenarios like the ones in OnnxConversionTests, where full pipelines are exported to onnx and applied to the same input, this is not much a problem. But in scenarios like the one in here, where a preexisting onnx model is applied as a featurizer, it might become a problem if the user wants to use again the "data_0" column.

Furthermore, this will be unexpected behavior, because the applied onnx model wouldn't have come from exporting a ColumnSelectingTransformer, but rather it might come from any other source. So dropping columns wouldn't be entirely right.

3 possible solutions I can think for this secnario:
1. The user should somehow know (perhaps through docs) that the OnnxTransformer will drop the input column, so if they want to use it they should copy the column before applying the model.
2. We could add Options to the ApplyOnnxModel method to let the user decide if they want to drop all the input columns or not, or even let them choose which one of them to drop. Then the GetDropColumnsNames in the OnnxTransformer would return the appropiate columns to drop.
3. We could save which columns to drop inside the Onnx model, in some sort of metadata (such as the SlotNames were saved on PR #4857 ) and then let the GetDropColumnsNames get those names from there.

Reply:
Another alternative that is occurring to me is that onnxtransformer can hide instead of drop the input columns. Is there a way for users to un-hide columns?

Reply:
So this version is better than the previous version of dropping every column except for the ONNX outputs, but I still think that it is introducing wrong behavior, because of the side effects that @antoniovs1029 mentioned above. Dropping the input columns will be wrong in most cases (most ONNX models won't contain `DropColumns`). In addition, this change might do the wrong thing even when the ONNX model **does** contain a `DropColumns` node, since there is no guarantee that the columns being dropped are the input columns, is there?

I still think that this change is solving a non-issue, and should be avoided. The expected behavior of the `OnnxTransformer` is to compute the column that the user requests, and it is not expected to change the rest of the data view schema.

If the only reason this issue has come up is that in the tests of the ONNX export of `ColumnSelectingTransformer` it seems as though the ONNX model is not doing what it is intended to do, I suggest testing this in a different way. One possibility, is to create an `OnnxTransformer` asking for one of the columns that have been dropped, and verifying that we get an error saying something like "there is no such column in the ONNX model" (this is probably not the syntax of the error message, but you get the idea). Again, I really think the change needs to be in the test, and not in the code of `OnnxTransformer`.

---
In reply to: [406110390](https://github.com/dotnet/machinelearning/pull/5012#discussion_r406110390) [](ancestors = 406110390)

Reply:
> In addition, this change might do the wrong thing even when the ONNX model does contain a DropColumns node, since there is no guarantee that the columns being dropped are the input columns, is there?

I am not sure I'm following this point,. "DropColumns nodes" don't exist in ONNX. since ONNX doesn't even have a notion of "column". The way the ColumnSelectingTransformer ONNX export works is that if a column was to be dropped, then the final ONNX model won't contain the output onnx node that would cause the column to propagate (this will happen whether the column is in the input of the pipeline being exported, or if the column is created inside the pipeline). The PR here will drop input columns that are dropped or hidden by the pipeline being exported, and it will also drop columns that were created inside the pipeline and also dropped or hidden inside the pipeline (this is demonstrated by cases 1a, 1b, and 3 of the ColumnSelectingOnnxTestColumnPropagation test I added, although I missed including the case of dropping columns that where created and hidden in the pipeline being exported).

So there's no guarantee that a SelectingColumnTransformer will drop columns from the input, since it could be dropping columns inside the pipeline being exported (it's either one or the other), and in the latter case the input column would have been hidden anyway. So, I think that the PR here works in the sense that it always drops columns that an exported SelectingColumnsTransformer will drop (although it also drops other hidden columns). On the other hand, the side effect I pointed to here is still a problem.

> I still think that this change is solving a non-issue, and should be avoided. The expected behavior of the OnnxTransformer is to compute the column that the user requests, and it is not expected to change the rest of the data view schema.

I've discussed this with @harishsk , and I also agree that this issue shouldn't be fixed. But he thinks otherwise, because if the SelectingColumnTransformer drops some columns, the onnx exported model should also drop all of them... so I agree this is a bug, but it's probably not one that should be fixed. Particularly, I think he wants this issue to be fixed, since it might cause performance issues in NimbusML, where when a user wants the outputs of a transforms pipeline, he receives the output of all unhidden columns... if the user has a pipeline that includes ColumnSelector to drop undesired columns, and then exports it to onnx, then it's the case that the exported version won't drop the input columns being propagated (which in some causes could be thousands of columns), and then when applying the onnx model, it takes a lot of time to retrieve the values of all the columns in the input.

Still, there's a workaround: let the user add a ColumnSelectingTransformer after applying the onnx model, and the problem is solved... I found that this works in both ML.NET and NimbusML. @harishsk is hesitant about this, since his opinion is that this should be done automatically since the user might not know which columns to drop. But unless we save some sort of metadata in the onnx model, to let ML.NET know how to drop columns from it, I don't see how can this be done without the side effect I pointed to here.

---
In reply to: [406669100](https://github.com/dotnet/machinelearning/pull/5012#discussion_r406669100) [](ancestors = 406669100,406110390)

Reply:
After discussing this offline with @yaeldekel and @harishsk , we've come to the conclussion that this side effect is not acceptable, and we should let users apply pretrained onnx models without dropping the input column.

So to fully fix this issue, there should be a way (perhaps metadata) to annotate in the onnx model which columns to drop, and then let GetDropColumnsNames to get those names from there.

---
In reply to: [406839974](https://github.com/dotnet/machinelearning/pull/5012#discussion_r406839974) [](ancestors = 406839974,406669100,406110390)